pkg/backup/scheduler.go

```diff
@@ -0,0 +1,141 @@
+package backup
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	logger "github.com/rs/zerolog/log"
+	"go.opentelemetry.io/otel/metric/global"
+	"go.opentelemetry.io/otel/metric/instrument"
+)
+
+var log = logger.With().Str("component", "backup").Logger()
+
+// Scheduler executes backups at a regular interval.
+type Scheduler struct {
+	NotificationCh chan error
+
+	notify          bool
+	backuper        *Backuper
+	tickerFrequency time.Duration
+
+	// control
+	close     chan struct{}
+	closeOnce sync.Once
+
+	// metrics
+	mLastExecution time.Time
+}
+
+// BackuperOptions options needed to instantiate a backuper.
+type BackuperOptions struct {
+	SourcePath, BackupDir string
+	Opts                  []Option
+}
+
+// NewScheduler creates a new backup scheduler.
+func NewScheduler(frequency int, opts BackuperOptions, notify bool) (*Scheduler, error) {
+	if frequency < 1 || frequency >= 1440 {
+		return nil, errors.New("frequency should be in [1,1440)")
+	}
+
+	backuper, err := NewBackuper(opts.SourcePath, opts.BackupDir, opts.Opts...)
+	if err != nil {
+		return nil, fmt.Errorf("new backuper: %s", err)
+	}
+
+	s := &Scheduler{
+		NotificationCh:  make(chan error),
+		notify:          notify,
+		backuper:        backuper,
+		tickerFrequency: time.Duration(frequency) * time.Minute,
```

Comment:
Using `tickerFrequency` as the only property to calculate how long to wait for the next backup run. In tests, we change this to a higher frequency to execute the test faster